lms/djangoapps/courseware/model_data.py

```diff
@@ -34,18 +55,652 @@ class InvalidWriteError(Exception):
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: bool
+        """
+        return self._cache_key_for_kvs_key(kvs_key) in self._cache
+
+    @contract(kvs_key=DjangoKeyValueStore.Key, returns="datetime|None")
+    def last_modified(self, kvs_key):
+        """
+        Return when the supplied field was changed.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: datetime if there was a modified date, or None otherwise
+        """
+        field_object = self._cache.get(self._cache_key_for_kvs_key(kvs_key))
+
+        if field_object is None:
+            return None
+        else:
+            return field_object.modified
+
+    def __len__(self):
+        return len(self._cache)
+
+    @abstractmethod
+    def _create_object(self, kvs_key, value):
+        """
+        Create a new object to add to the cache (which should record
+        the specified field ``value`` for the field identified by
+        ``kvs_key``).
+
+        Arguments:
+            kvs_key (:class:`DjangoKeyValueStore.Key`): Which field to create an entry for
+            value: What value to record in the field
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _read_objects(self, fields, xblocks, aside_types):
+        """
+        Return an iterator for all objects stored in the underlying datastore
+        for the ``fields`` on the ``xblocks`` and the ``aside_types`` associated
+        with them.
+
+        Arguments:
+            fields (list of str): Field names to return values for
+            xblocks (list of :class:`~XBlock`): XBlocks to load fields for
+            aside_types (list of str): Asides to load field for (which annotate the supplied
+                xblocks).
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _cache_key_for_field_object(self, field_object):
+        """
+        Return the key used in this DjangoOrmFieldCache to store the specified field_object.
+
+        Arguments:
+            field_object: A Django model instance that stores the data for fields in this cache
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _cache_key_for_kvs_key(self, key):
+        """
+        Return the key used in this DjangoOrmFieldCache for the specified KeyValueStore key.
+
+        Arguments:
+            key (:class:`~DjangoKeyValueStore.Key`): The key representing the cached field
+        """
+        raise NotImplementedError()
+
+
+class UserStateCache(object):
     """
-    Yields the values from items in chunks of size chunk_size
+    Cache for Scope.user_state xblock field data.
     """
-    items = list(items)
-    return (items[i:i + chunk_size] for i in xrange(0, len(items), chunk_size))
+    def __init__(self, user, course_id):
+        self._cache = defaultdict(dict)
```

Comment:
Does `self._cache` grow without bound? If so, will this potentially be a problem?